src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -3020,19 +3020,45 @@ ReceiveResults(WorkerSession *session, bool storeRows)
 	uint32 expectedColumnCount = 0;
 	char **columnArray = NULL;
 	Tuplestorestate *tupleStore = execution->tupleStore;
+	MemoryContext oldContext = NULL;
+	MemoryContext ioContext = NULL;
+
+	MemoryContext receiveResultsContext = AllocSetContextCreate(CurrentMemoryContext,
+																"ReceiveResults",
+																ALLOCSET_DEFAULT_MINSIZE,
+																ALLOCSET_DEFAULT_INITSIZE,
+																ALLOCSET_DEFAULT_MAXSIZE);
+
+
+	/*
+	 * Allocate all memory within this context. This function may be called many times
+	 * per distributed execution, so we're guarding against any memory leaks with using a
+	 * specific memory context. This context is deleted at the end of the function, so
+	 * be careful to add early returns to the function.
+	 *
+	 * Finally, tuplestore_puttuple() copies the tuples that are stored in the tuplestore
+	 * in the tuplestore's own memory context. So, we don't need to anything special
+	 * for that.
+	 */
+	oldContext = MemoryContextSwitchTo(receiveResultsContext);
 
-	MemoryContext ioContext = AllocSetContextCreate(CurrentMemoryContext,
-													"ReceiveResults",
-													ALLOCSET_DEFAULT_MINSIZE,
-													ALLOCSET_DEFAULT_INITSIZE,
-													ALLOCSET_DEFAULT_MAXSIZE);
 	if (tupleDescriptor != NULL)
 	{
 		attributeInputMetadata = TupleDescGetAttInMetadata(tupleDescriptor);
 		expectedColumnCount = tupleDescriptor->natts;
 		columnArray = (char **) palloc0(expectedColumnCount * sizeof(char *));
```

Comment:
could we just store these in the DistributedExecution to avoid re-allocating them at all?

Reply:
makes sense, I added it. Well, the extra memory context (`ReceiveResults`) now seems useless. Still, I've a slight preference to keep it to prevent such cases all together. What is your take?